File: impl/go/internal/chanlun/fractal.go
package chanlun

// Def-3 (Fractal) classifier, faithful port of
// conformance/chanlun-v1/reference_backend/fractal.py.
//
// Top at i:    H_i > H_{i-1} && H_i > H_{i+1} && L_i > L_{i-1} && L_i > L_{i+1}
// Bottom at i: L_i < L_{i-1} && L_i < L_{i+1} && H_i < H_{i-1} && H_i < H_{i+1}
// Neither is the named structural residue and is dropped (gaps in idx).

const (
	KindTop    = "top"
	KindBottom = "bottom"
)

// Fractal is an emitted (top or bottom) fractal at position idx in the bar list.
type Fractal struct {
	Idx  int64
	Kind string
	H    int64
	L    int64
}

func isTop(a, b, c Bar) bool {
	return b.H > a.H && b.H > c.H && b.L > a.L && b.L > c.L
}

func isBottom(a, b, c Bar) bool {
	return b.L < a.L && b.L < c.L && b.H < a.H && b.H < c.H
}

// ExtractFractals scans an already-normalized bar list and emits the
// {idx, kind, h, l} sequence (interior windows only).
func ExtractFractals(bars []Bar) []Fractal {
	n := len(bars)
	out := make([]Fractal, 0)
	for i := 1; i < n-1; i++ {
		a, b, c := bars[i-1], bars[i], bars[i+1]
		switch {
		case isTop(a, b, c):
			out = append(out, Fractal{Idx: int64(i), Kind: KindTop, H: b.H, L: b.L})
		case isBottom(a, b, c):
			out = append(out, Fractal{Idx: int64(i), Kind: KindBottom, H: b.H, L: b.L})
		default:
			// 'neither' - dropped (named residue, gap in idx)
		}
	}
	return out
}

// FractalsFromValue parses a JSON list of fractals.
func FractalsFromValue(v Value) []Fractal {
	arr := v.MustArr()
	out := make([]Fractal, len(arr))
	for i, e := range arr {
		o := e.MustObj()
		idx, _ := o.Get("idx")
		kind, _ := o.Get("kind")
		h, _ := o.Get("h")
		l, _ := o.Get("l")
		out[i] = Fractal{
			Idx:  idx.MustInt(),
			Kind: kind.MustStr(),
			H:    h.MustInt(),
			L:    l.MustInt(),
		}
	}
	return out
}

// FractalsToValue emits the fractal list as a typed Value tree.
func FractalsToValue(fs []Fractal) Value {
	out := make([]Value, len(fs))
	for i, f := range fs {
		o := NewObject()
		o.Set("idx", VInt(f.Idx))
		o.Set("kind", VStr(f.Kind))
		o.Set("h", VInt(f.H))
		o.Set("l", VInt(f.L))
		out[i] = VObj(o)
	}
	return VArr(out)
}

File: impl/csharp/src/Chanlun/Fractal.cs
// Def-3 fractal classifier — faithful port of
// conformance/chanlun-v1/reference_backend/fractal.py.
//
//   Top at i:    H_i > H_{i-1} && H_i > H_{i+1} && L_i > L_{i-1} && L_i > L_{i+1}
//   Bottom at i: L_i < L_{i-1} && L_i < L_{i+1} && H_i < H_{i-1} && H_i < H_{i+1}
// Neither = named structural residue (dropped; gaps in idx).

namespace Chanlun;

public static class FractalOps
{
    public static List<Fractal> ExtractFractals(IReadOnlyList<Bar> bars)
    {
        var out_ = new List<Fractal>();
        int n = bars.Count;
        for (int i = 1; i < n - 1; i++)
        {
            var a = bars[i - 1];
            var b = bars[i];
            var c = bars[i + 1];
            if (IsTop(a, b, c))
                out_.Add(new Fractal(i, Kinds.Top, b.H, b.L));
            else if (IsBottom(a, b, c))
                out_.Add(new Fractal(i, Kinds.Bottom, b.H, b.L));
            // neither -> dropped
        }
        return out_;
    }

    private static bool IsTop(Bar a, Bar b, Bar c)
        => b.H > a.H && b.H > c.H && b.L > a.L && b.L > c.L;

    private static bool IsBottom(Bar a, Bar b, Bar c)
        => b.L < a.L && b.L < c.L && b.H < a.H && b.H < c.H;

    // ---- Value-tree converters ------------------------------------------------

    public static List<Fractal> FractalsFromValue(Value v)
    {
        var arr = v.AsArray();
        var out_ = new List<Fractal>(arr.Count);
        foreach (var e in arr)
        {
            var o = e.AsObject();
            out_.Add(new Fractal(
                o.Get("idx").AsInt(),
                o.Get("kind").AsString(),
                o.Get("h").AsInt(),
                o.Get("l").AsInt()));
        }
        return out_;
    }

    public static Value FractalsToValue(IReadOnlyList<Fractal> fs)
    {
        var arr = new List<Value>(fs.Count);
        foreach (var f in fs)
        {
            var o = new OrderedObject();
            o.Set("idx", Value.Of(f.Idx));
            o.Set("kind", Value.Of(f.Kind));
            o.Set("h", Value.Of(f.H));
            o.Set("l", Value.Of(f.L));
            arr.Add(Value.Of(o));
        }
        return Value.Of(arr);
    }
}
